### ISSUE TITLE
Question about patch selection code

### ISSUE THREAD

**Issue by WangYijun-OUC** [Hanminghao/MSCPT#10]
Thanks for your work! In file [select_5X_pic](https://github.com/Hanminghao/MSCPT/blob/master/select_5X_pic.py), I see you compute the similarities between patches and prompts in all categories(like I use dataset for 30 classes) in 
```python
logits = text_feats @ img_feats.T
logits = logits.cpu()
```
And the shape of logits is _[30, patches]_. While in code
```python
if args.top_k > img_feats.shape[0]:
    topk_values, topk_indices = torch.topk(logits, img_feats.shape[0], dim=1)
else:
    topk_values, topk_indices = torch.topk(logits, args.top_k, dim=1)
pred = topk_values.sum(dim=1).argmax().cpu().item()
select_id = topk_indices.flatten().cpu().numpy()
coord = coords[select_id]
scores = logits[pred][select_id].numpy()
df.loc[idx, 'pred'] = 1 if pred == category else 0  # 进行赋值
for idx, (x,y) in enumerate(coord):
    big_img = wsi.read_region((x,y), patch_level, (patch_size, patch_size)).convert('RGB')
    big_img = big_img.resize((224,224))
    big_img.save(os.path.join(save_path, f"score_{scores[idx]:.4f}_{idx}_{x}_{y}.png"))
```
I see you compute the top-k similarities for each class, and _select_id_ is the number of all patches in all classes, like _30 * 30 = 900_ patches. But as you already compute the pred class, do we really need all the patches in all classes including the pred class? Because I see the image saving results _big_img_ contains a lot of overlapping patches, which have the save _score_, _x and y coordinates_，but different _idx_. 

![Image](https://github.com/user-attachments/assets/a789b222-61d9-487a-b347-0cbc3e9c37a8)

I feel strange about it, wish you could help me.

**Comment by Hanminghao** [comment]
Thank you for your attention. In this case, the "pred" classes are selected using the zero-shot capabilities of the VLM and may not be entirely accurate. Therefore, during the subsequent few-shot training process, we require more comprehensive information rather than relying solely on data from the "assumed" classes. Consequently, we aim to retain all big_img instances associated with all potential classes.

**Comment by WangYijun-OUC** [comment]
Thank you for your response. In my own dataset, however, almost all of the “pred” values in the final CSV end up as 0.0. Could this be due to the quality of the textual prompts, and affect the training process?

Also, I have a follow‑up: what is the difference between the prompts used in the patch‑selection step (see [generate_select_prompt.py](https://github.com/Hanminghao/MSCPT/blob/master/generate_select_prompt.py)) and those used during model training (see [BRCA.json.py](https://github.com/Hanminghao/MSCPT/blob/master/train_data/gpt/description/BRCA.json))? Were both sets generated by LLM such as GPT‑4? If so, could you share the exact LLM query you used to produce the patch selection prompt?

**Comment by WangYijun-OUC** [comment]
One last question: did you employ a distributed training strategy? It seems like model is trained on single GPU.

**Comment by Hanminghao** [comment]
Thank you for your attention.

A1: I believe this is influenced by both the quality of the textual prompts and the performance of the VLM. You could try using CONCH as the base VLM and employing higher-quality textual descriptions.

A2: The prompt used in the patch-selection step is a relatively coarse prompt, only used for the initial selection of patches that will participate in downstream visual prompt tuning. It does not contain specific visual-language descriptions of cancer but is simply "This is a XXXX type of cancer." The prompts used in model training, however, contain knowledge-intensive content, such as visual-language descriptions of different types of cancer at various scales. We mentioned the specific LLM query instructions in Section III.C.

A3: Since we are doing prompt tuning, the computational resource requirements are not very high, so we did not use distributed training. However, as the number of patches involved in visual prompt tuning increases, the demand for VRAM will significantly rise. You mentioned earlier that you have 30 categories and select 30 patches per category, meaning 900 patches participate in model training. I am concerned about the VRAM requirements in this scenario. If I recall correctly, the VRAM requirement for 900 patches is far greater than 80GB.

**Comment by WangYijun-OUC** [comment]
Thanks for your answer, I will soon close this issue.
